src/highlights/highlight_grid.ts

```diff
@@ -1,152 +1,323 @@
-import wcswidth from "ts-wcwidth";
-import GraphemeSplitter from "grapheme-splitter";
-
-import { calculateEditorColFromVimScreenCol, expandTabs } from "../utils/text";
-
-export interface ValidCell {
+import { cloneDeep } from "lodash";
+import {
+    Range,
+    ThemeColor,
+    type DecorationOptions,
+    type Disposable,
+    type TextEditor,
+    type TextEditorDecorationType,
+} from "vscode";
+
+import { BufferManager } from "../buffer_manager";
+import { ViewportManager, type Viewport } from "../viewport_manager";
+
+import { HighlightGroupStore } from "./highlight_group_store";
+import { CellIter, getWidth, isDouble, splitGraphemes } from "./util";
+
+export type VimCell = [string, number?, number?];
+export interface LineCell {
     text: string;
     hlId: number;
 }
-
-export interface Highlight extends ValidCell {
+export interface Highlight {
+    text: string;
+    hlId: number;
     virtText?: string;
 }
-
-export type HighlightRange = NormalTextHighlightRange | VirtualTextHighlightRange;
-
-export interface NormalTextHighlightRange {
+export interface NormalHighlightRange {
     textType: "normal";
     hlId: number;
     line: number;
     startCol: number;
     endCol: number;
 }
-
-export interface VirtualTextHighlightRange {
+export interface VirtualHighlightRange {
     textType: "virtual";
     highlights: Highlight[];
     line: number;
     col: number;
 }
+export type HighlightRange = NormalHighlightRange | VirtualHighlightRange;
+
+export class HighlightGrid implements Disposable {
+    // line number -> line cells
+    private lineCells: LineCell[][] = [];
+    // The way to clear decorations is to set them to an empty array, so it is
+    // necessary to record the decorators used in the last refresh.
+    // In the next refresh, if a decorator is no longer used, it should be cleared.
+    private prevDecorators: Set<TextEditorDecorationType> = new Set();
+    // Flag to indicate if the grid needs to be redrawn
+    private isDirty = false;
+
+    // Used to get the editor and viewport
+    private readonly gridId: number;
+    // Normalizes highlight IDs and provides decorators
+    private readonly groupStore: HighlightGroupStore;
+    // Not available when testing, so optional
```

Comment:
This, in particular, I think is a smell we should address. `HighlightGrid` was able to be fully unit tested before, because it didn't have any dependencies on our VSCode state; it was passed in as needed. As far as I can tell, these two properties are only necessary because decorations have been merged into `HighlightGrid`, which is a responsibility we can delegate either to the `Manager` or the (now removed) `Provider` like before.

Reply:
Alright, If it weren't for testing purposes, the code could be further simplified a lot.

I'm thinking of splitting it into separate files, like calculations.ts. What do you think?

Reply:
:100: 

I think we're on the same page. This is what I had hoped to do with `HighlightGrid` originally (it was a data structure which could maintain our highlight calculations, and could then be used by the `Provider` to produce decorations), and I think if we continue to maintain that separation, we will avoid complexity to just facilitate testing.

BTW, I do intend to review more than just the testing logic, I just haven't gotten to it yet :) I'm very interested in this approach

Reply:
Splitting the logic for testing purposes has ironically made the logic more complicated, and I'm really conflicted about it.

Reply:
It's a give and take, for sure. I'm confident you'll find a good middle ground :)

Reply:
I'm doing something I don't like doing.

Reply:
If there's any way I can help, do let me know